core/src/main/java/org/infinispan/transaction/impl/TransactionCoordinator.java

```diff
@@ -99,92 +107,116 @@ public PrepareCommand createPrepareCommand(GlobalTransaction gtx, List<WriteComm
+               }
+               return CompletableFutures.completedExceptionFuture(xe);
+            });
+         }
          if (localTransaction.isReadOnly()) {
             if (trace) log.tracef("Readonly transaction: %s", localTransaction.getGlobalTransaction());
             // force a cleanup to release any objects held.  Some TMs don't call commit if it is a READ ONLY tx.  See ISPN-845
-            commitInternal(ctx);
-            return XA_RDONLY;
+            return commitInternal(ctx)
+                  .thenApply(XA_RDONLY_APPLY);
          } else {
             txTable.running().localTransactionPrepared(localTransaction);
-            return XA_OK;
+            return XA_OKAY_STAGE;
          }
-      } catch (Throwable e) {
-         if (shuttingDown)
-            log.trace("Exception while preparing back, probably because we're shutting down.");
-         else
-            log.errorProcessingPrepare(e);
-
-         //rollback transaction before throwing the exception as there's no guarantee the TM calls XAResource.rollback
-         //after prepare failed.
-         rollback(localTransaction);
-         // XA_RBROLLBACK tells the TM that we've rolled back already: the TM shouldn't call rollback after this.
-         XAException xe = new XAException(XAException.XA_RBROLLBACK);
-         xe.initCause(e);
-         throw xe;
-      }
+      });
    }
 
-   public boolean commit(LocalTransaction localTransaction, boolean isOnePhase) throws XAException {
+   public CompletionStage<Boolean> commit(LocalTransaction localTransaction, boolean isOnePhase) {
       if (trace) log.tracef("Committing transaction %s", localTransaction.getGlobalTransaction());
       LocalTxInvocationContext ctx = icf.running().createTxInvocationContext(localTransaction);
       if (isOnePhaseCommit(localTransaction) || isOnePhase) {
-         validateNotMarkedForRollback(localTransaction);
+         CompletionStage<Boolean> markRollbackStage = validateNotMarkedForRollback(localTransaction);
+         if (markRollbackStage != null) {
+            return markRollbackStage;
+         }
 
          if (trace) log.trace("Doing an 1PC prepare call on the interceptor chain");
          List<WriteCommand> modifications = localTransaction.getModifications();
          PrepareCommand command = commandCreator.createPrepareCommand(localTransaction.getGlobalTransaction(), modifications, true);
          try {
-            invoker.running().invoke(ctx, command);
+            return CompletionStages.handleAndCompose(invoker.running().invokeAsync(ctx, command),
+                  (ignore, t) -> {
+                     if (t != null) {
+                        return handleCommitFailure(t, true, ctx);
+                     }
+                     return CompletableFutures.completedTrue();
+                  });
          } catch (Throwable e) {
-            handleCommitFailure(e, true, ctx);
+            return handleCommitFailure(e, true, ctx);
```

Comment:
No need for try..catch

Reply:
+1, yeah I didn't go back and double check I removed all of them.